### PR TITLE
rbd: support namespace. (Step1)

### PR DESCRIPTION
**Pull request by yangdongsheng**
Fix: http://tracker.ceph.com/issues/19073

This is Step 1 for rbd namespace, then we can use rbd namespace as below. 

```
# rbd ls --all --long
# rbd create -N test_ns test -s 1
# rbd ls 
# rbd -N test_ns ls
test
# rbd ls --all --long
NAME  SIZE PARENT FMT PROT LOCK NAMESPACE 
test 1024k          2           test_ns   
```

### CODE REVIEW

**Comment by yangdongsheng** [comment]
@dillaman I found unittest does not support nspace now. will investigate it more. In addition, I am planing to use omaps of a rbd_namespaces in default namespace to list the all images in a pool.

**Comment by dmick** [comment]
My fault the submodule test failed, ignore

**Comment by yangdongsheng** [comment]
@dmick thanx for the clarify. @ceph-jenkins retest please

**Comment by dmick** [comment]
retest this please

**Comment by yangdongsheng** [comment]
Need some more consideration, will reopen later.

**Comment by yangdongsheng** [comment]
rbd_namespace

Hi @dillaman There is an idea about namespace supporting, as shown above. Considering the backward compatibility, I will not change anything in default namespace, that means, for the old version of librbd, it should not notice anything different. But for the newer version of librbd, it can see the view in this picture.

(1) global_ns: it's a special namespace in a pool, we store some global information in this namespace. such as indexing of namespace, rbd_namespace. 
(2) For the rbd mirroring, I think we have to store some information in the global namespace, such as mirror mode. When we create an image in a namespace, we need to get mirror mode from global_ns, if the mode is pool, we should enable the mirroring for this image. And each namespace will have their own rbd_mirroring for themselves, as what is looks like as before, but this rbd_mirroring only register the images in their own namespace. 
(3) default namespace: the default namespace will be nothing different with what it was. 
(4) ns_xxx: but the new namespaces will be prefixed with ns_. This is used to prevent user accessing global_ns.

Please help to take a look and give me some suggestion about it. The current implementation in this PR is not for this design. But I will update this PR as what shown in the picture.

**Comment by dillaman** [comment]
@yangdongsheng Would there be much difference between the "(empty)" namespace and "global_ns"? If you are using namespaces for isolation, you would only give the user access to their namespace and perhaps a "write+execute" access cap for a cls_method on rbd_namespace object in the "(empty)" namespace (to allow a client register the unique namespace if desired). We would also need some sort of state guard to prevent the removal of the namespace registration while the namespace still has images associated into the pool.

**Comment by yangdongsheng** [comment]
@dillaman thanx for your reply. yes, after a second thought, the global_ns seems not needed, just using the "" (empty) default namespace. then the overview of the namespace of a pool would be that:
rbd_namespace_1

rbd_namespace is the indexing of the all namespaces in this pool.

But we have to do some more work about rbd_mirror. if we create a new image in a new namespace, we have to check the mirroring mode in default namespace, if that's POOL. we have to enable the mirroring in the new namespace. like that:
rbd_namespace_2



At the same time, we have to change the behavior of the rbd mirror pool command. such as rbd mirror pool enable.
rbd_namespace_3



Then when we set the rbd mirror mode to POOL, the all images in each namespace would be mirrored with peers.  Although we have to change the rbd_mirror tool to handle this case, maybe we should rename Replayer in rbd_mirror to NSReplayer. and add a new PoolReplayer to container many different NSReplayers. 

In addition, we will support mirroring mode of NAMESPACE to make a namespace mirrored with namespace. 

I will continue this design and implement it if you think that would working. :)

**Comment by yangdongsheng** [comment]
@dillaman wait a minute, I think I am making this work too complex. Maybe we should only isolate a pool by namespace currently. So let me try a much simpler solution: Just isolate everything in a pool by namespace, as below:
```
$ rbd -c local.conf info test
rbd image 'test':
	size 1024 MB in 256 objects
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.102f2ae8944a
	format: 2
	features: layering, exclusive-lock, object-map, fast-diff, deep-flatten, journaling
	flags: 
	journal: 102f2ae8944a
	mirroring state: enabled
	mirroring global id: 9b300e22-bfa6-48e3-b651-f01de07884d0
	mirroring primary: true
$ rbd -c local.conf mirror pool info rbd
Mode: pool
Peers: 
  UUID                                 NAME   CLIENT       
  37bba207-cd76-4090-af6a-4d27445fd337 remote client.admin 
$ rbd -c local.conf mirror pool -N test info rbd
Mode: disabled
$ rbd -c local.conf create -N test_ns test -s 1
$ rbd -c local.conf mirror pool -N test info rbd
Mode: disabled
$ rbd -c local.conf -N test_ns info test
rbd image 'test':
	size 1024 kB in 1 objects
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.8561238e1f29
	format: 2
	features: layering, exclusive-lock, object-map, fast-diff, deep-flatten, journaling
	flags: 
	journal: 8561238e1f29
	mirroring state: disabled
$ rbd -c local.conf -N "" info test
rbd image 'test':
	size 1024 MB in 256 objects
	order 22 (4096 kB objects)
	block_name_prefix: rbd_data.102f2ae8944a
	format: 2
	features: layering, exclusive-lock, object-map, fast-diff, deep-flatten, journaling
	flags: 
	journal: 102f2ae8944a
	mirroring state: enabled
	mirroring global id: 9b300e22-bfa6-48e3-b651-f01de07884d0
	mirroring primary: true
$ rbd -c local.conf ls --all --long
NAME  SIZE PARENT FMT PROT LOCK NAMESPACE 
test 1024M          2           default   
test 1024k          2           test_ns  
```
Please help to take a look at the current implementation. thanx

**Comment by yangdongsheng** [comment]
@ceph-jenkins  retest this please

File: src/cls/rbd/cls_rbd.cc [hunk 1]
```diff
@@ -2784,6 +2803,137 @@ int snapshot_set_limit(cls_method_context_t hctx, bufferlist *in,
   return rc;
 }
 
+int get_namespace(cls_method_context_t hctx, bufferlist *in,
+		  bufferlist *out)
+{
+  int rc;
```
Comment (amitkumar50) [inline src/cls/rbd/cls_rbd.cc]: @yangdongsheng How about using uint32_t or int32_t?

**Comment by yangdongsheng** [comment]
@PCzhangPC  will take over this work and open another PR for rbd namespace. So close this one now.

**Pull request closed without merging**